File: JavaPractice/src/ATM.java
/**
 * @author divyesh.shah
 */
import java.util.Arrays;
/*
* Withdraw bills from ATM or dispense money bill or cash
* Find number of counts
* */
public class ATM {
    public static int[] withdrawAmount(int[] billsAvailable, int desiredAmount) {
        if (desiredAmount % 10 != 0) {
            System.out.println("Desired amount must be a multiple of 10.");
            return new int[0];  // Return an empty array
        }

        int[] sortedBills = Arrays.stream(billsAvailable)
                .boxed()
                .sorted((bill1, bill2) -> Integer.compare(bill2, bill1))
                .mapToInt(Integer::intValue)
                .toArray();

        int[] withdrawBills = new int[sortedBills.length];
        int remainingAmount = desiredAmount;

        for (int i = 0; i < sortedBills.length; i++) {
            int billCount = remainingAmount / sortedBills[i];
            withdrawBills[i] = billCount;
            remainingAmount -= sortedBills[i] * billCount;
        }

        if (remainingAmount != 0) {
            System.out.println("Unable to dispense the desired amount with available bills.");
            return new int[0];  // Return an empty array
        }

        return withdrawBills;
    }

    public static void main(String[] args) {
        int[] billsAvailable = {10, 40, 20, 50};
        int desiredAmount = 330;

        int[] withdrawBills = withdrawAmount(billsAvailable,desiredAmount);
        int count = 0;
        if (withdrawBills.length > 0) {
            System.out.println("Bills returned:");
            for (int i = 0; i < withdrawBills.length; i++) {
                if (withdrawBills[i] > 0) {
                    System.out.println(billsAvailable[i] + " * " + withdrawBills[i]);
                    count += withdrawBills[i];
                }
            }
        }
        System.out.println("No of bills - " + count);
    }
}

File: PythonPractice/src/ATM.py
from typing import List, Optional

def withdraw_amount(bills_available: List[int], desired_amount: int) -> Optional[List[int]]:
    # Check if the desired amount is a multiple of 10
    if desired_amount % 10 != 0:
        print("Desired amount must be a multiple of 10.")
        return []  # Return an empty list

    # Sort the bills in descending order
    sorted_bills = sorted(bills_available, reverse=True)

    # Initialize the list to hold the number of each bill to be withdrawn
    withdraw_bills = [0] * len(sorted_bills)
    remaining_amount = desired_amount

    # Calculate the number of each bill to withdraw
    for i, bill in enumerate(sorted_bills):
        if bill <= 0:
            print(f"Invalid bill denomination: {bill}.")
            return []  # Return an empty list

        bill_count = remaining_amount // bill
        withdraw_bills[i] = bill_count
        remaining_amount -= bill * bill_count

    # Check if the remaining amount is zero
    if remaining_amount != 0:
        print("Unable to dispense the desired amount with available bills.")
        return []  # Return an empty list

    return withdraw_bills

def main():
    bills_available = [10, 40, 20, 50]
    desired_amount = 330

    withdraw_bills = withdraw_amount(bills_available, desired_amount)
    count = 0

    if withdraw_bills:
        print("Bills returned:")
        for i, bill_count in enumerate(withdraw_bills):
            if bill_count > 0:
                print(f"{bills_available[i]} * {bill_count}")
                count += bill_count

    print(f"No of bills - {count}")

if __name__ == "__main__":
    main()
